### PR TITLE
persist: properly enforce uniqueness in futures/traces

### PR DESCRIPTION
**Pull request by ruchirK**
Noticed this today while I was poking around related code

### CODE REVIEW

**Review by danhhz — APPROVED** [review]
LGTM assuming you're cool with these two changes

File: src/persist/src/indexed/encoding.rs [hunk 1]
```diff
@@ -206,8 +206,26 @@ impl BlobMeta {
             }
             t.validate()?;
         }
-        let futures: HashMap<_, _> = self.futures.iter().map(|(id, m)| (*id, m)).collect();
-        let traces: HashMap<_, _> = self.traces.iter().map(|(id, m)| (*id, m)).collect();
+        let mut futures = HashMap::new();
```
[the remark this answers is not in the document]
Comment (ruchirK) [reply src/persist/src/indexed/encoding.rs:209]: hmm it looks like we end up using the meta field in both cases a little down the line in lines 230 / 233?

File: src/persist/src/indexed/encoding.rs [hunk 2]
```diff
@@ -206,8 +206,26 @@ impl BlobMeta {
...
+
+        for (id, meta) in self.futures.iter() {
```
[the remark this answers is not in the document]
Comment (ruchirK) [reply src/persist/src/indexed/encoding.rs]: done

File: src/persist/src/indexed/encoding.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist/src/indexed/encoding.rs]: Oh! Yeah ignore this. I missed the bit where these were here before

File: src/persist/src/indexed/encoding.rs [hunk 3]
```diff
@@ -194,20 +194,35 @@ impl BlobMeta {
...
             if !ids.contains(id) {
                 return Err(format!("futures id {:?} not present in id_mapping", id).into());
             }
+            
+            if futures.contains_key(id) {
+                return Err(format!("duplicate future: {:?}", id).into());
+            }
+
             f.validate()?;
+            futures.insert(*id, f);
```
Comment (danhhz) [inline src/persist/src/indexed/encoding.rs:209]: super nit: insert this above the valiate call to keep it next to the contains check

Comment (ruchirK) [reply src/persist/src/indexed/encoding.rs]: done!

**Comment by ruchirK** [comment]
thanks for the review! merging on green :)

**Pull request merged**